### DISCUSSION TITLE
Exclude Pattern for Cache Directories

### DISCUSSION THREAD

**Discussion by slauzon** [garethgeorge/backrest#469] [Q&A]
I'm attempting to exclude cache directories/files from my home directory backups on a windows PC.  I created a backup plan with the exclude (tried both case-sensitive and case-insensitive) field containing "`**cache** **cache2**`".  However, the backup keeps attempting to backup cache files and directories (example below).  Any help is greatly appreciated.

`
Error on item: C:\Users\<readacted>\AppData\Local\AMD\DxCache\b51bb678dd41a27.36b2c627f214d906.7b1f799a.dad1858.0.parc
Error on item: C:\Users\<readacted>\AppData\Local\BraveSoftware\Brave-Browser\User Data\Default\Cache\Cache_Data\data_0
`

**Comment by garethgeorge** [comment]
My iexcludes are set as `*cache*`, `*temp*`, and `*tmp*` . I'm not sure if restic (used under the hood) will handle the double `**` correctly -- you can take a look at https://restic.readthedocs.io/en/latest/040_backup.html#excluding-files for details on how those options are processed.

**Reply by slauzon** [reply]
Hi.  Thanks for responding.  I changed my iexcludes field to `*cache* *temp* *tmp*` as you suggested, but it appears that it's still attempting to backup directories it shouldn't.  See error below.

`[restic] {"message_type":"verbose_[tasklog] 2024-09-14T14:11:31.997-0500	warn	backup for plan "desktop"	an unknown error was encountered in processing item: C:\Users\<redacted>\AppData\Local\BraveSoftware\Brave-Browser\User Data\Default\Cache\Cache_Data\data_0`

**Comment by amarendra** [comment]
I kind of have a related doubt:

Can I just provide a file with list of paths/files to backup and a file with list of paths/files to ignore/exclude?

In restic cli I used to achieve that by exporting:
```
export RESTIC_INCLUDE_FILE="$HOME/path/restic.include"
export RESTIC_EXCLUDE_FILE="$HOME/path/restic.exclude"
```

and then using them as 
```
	--files-from $RESTIC_INCLUDE_FILE \
	--exclude-caches \
	--exclude-file=$RESTIC_EXCLUDE_FILE >> $RESTIC_BACKUP_LOG_FILE
```

But in Backrest all I am seeing is "Paths" and apparently I have to add every path one by one - is that correct? Anyway use the include file or add some other config file where I can keep that list instead of adding one by one.

**Reply by amarendra** [reply]
(Update) I am trying this:

In Repo (b2) - assuming these env vars will be available to plans as well:

```
  "env": [
    "B2_ACCOUNT_ID=----",
    "B2_ACCOUNT_KEY=----",
    "RESTIC_EXCLUDE_FILE=/Users/---/restic/restic.exclude",
    "RESTIC_INCLUDE_FILE=/Users/---/restic/restic.include",
    "RESTIC_PASSWORD_FILE=/Users/---/restic/.restic.pwd"
  ],

```
and then in my only plan I have tried to read as:

```
  },
  "backup_flags": [
    "--files-from=$RESTIC_INCLUDE_FILE",
    "--exclude-caches",
    "--exclude-file=$RESTIC_EXCLUDE_FILE",
    "--verbose"
  ],
```

Hopefully this works. It's failing right now. Trying to get some log or so. (Though `RESTIC_PASSWORD_FILE` has been getting read properly and I was abe to see list of snapshots - so might be able to fix it)

**Reply by garethgeorge** [reply]
Hey -- so a few behaviors to be aware of 

 - Backrest doesn't do env var substitution in CLI flags actually, it's something I could look at (probably worth a feature request if this is something you're interested in).
 - Backrest does do substitution in the env vars BUT the format must be `${MY_ENV_VAR}` note that the { } brackets denoting the name of the env var are important. 

password files , exclude files, exclude files, etc should all be passed through properly.

The last thing I'll mention, backrest does forward the environment passed to backrest. So if you're using docker or a wrapper script, you can set the env variables on the backrest process and they will be inherited by all commands.

**Reply by amarendra** [reply]
Hey thanks for getting back.

> worth a feature request

1. I mean it's fine either way. I will have only one Restic setup. So either it will be via Backrest, cli restic, or some other GUI (there's nothing else decent right now and CLI restic needed lots of bells and whistles done with bare hands)
1.1. But can there be a feature added where I can specify the location of those JSON config files and then on a new machie/reinstall I can just import those JSON files.
1.2. Because that is the whole point of these include and exclude files and env var configs - not having to type all over every time.
1.3 And just to confirm - so as of now I will have to enter paths of all the files/folders I want to include in backup and exclude I have to enter one by one using `Path` and `Excludes`. Right?

> Backrest does do substitution in the env vars BUT the format must be ${MY_ENV_VAR}

2. So you mean if I need to tell backrest my home directory then I will have to set `"RESTIC_PASSWORD_FILE=${$HOME}/config/restic/.restic.pwd"` or just `"RESTIC_PASSWORD_FILE=${HOME}/config/restic/.restic.pwd"` if `$HOME` is available on my laptop?
2.1 Will only ever use it on a personal laptop
2.2. I had [this](https://github.com/creativeprojects/resticprofile/discussions/273#discussioncomment-7399103) setup really (By the way, I could never get it working with `resticprofiles` from that post, so I had given up and explored Kopia etc but nothing stuck and I left my so far backed up data in b2. Seeing backrest an its clean, I thought  I will give it another try).

3. Is there any screencast/screenshots or detailed example tutorial where actual values/flags (kinda in detail) are shown as they are used so that one can do the first setup without stumbling around so much? Often `man` inspired doc might not help everybody much.

4. Also, are there options to add flags to prune, stats, check etc - if needed?

5. Any way to set log output file(s) locations? Right now I do not know where to find logs.

**Reply by garethgeorge** [reply]
1. Not quite sure if I follow what you're asking here -- I have some multihost management features planned (still prototyping in this space) that might address some of the pain points you're seeing here. In the mean time, I think what you're looking at re: using an exclude list as a file is a good approach.
2. Right -- but the format there should be the second e.g. `RESTIC_PASSWORD_FILE=${HOME}/config/restic/.restic.pwd`. $HOME should be defined on anything unix.
3. Afraid not though I'd love to get something like this put together. The most detailed docs I have are probably here: https://garethgeorge.github.io/backrest/introduction/getting-started . Most flags and env vars map directly to the same options in restic though, backrest aims to be a rather thin wrapper.
4. Flags on prune and check aren't supported; there aren't very many options offered on those commands and I think there are more footguns than not in exposing them but if there's a compelling use case I'll definitely consider it.
5. The log location is printed when backrest starts up -- or if it's running in the background this is usually in the data directory e.g. `$HOME/.local/share/backrest/processlogs/` on linux.

**Reply by amarendra** [reply]
1. By that meant kind of import/export of settings/config which takes care of everything including the choices I had made on previous system for config and file selection or exclusion etc. Let's say I can export a combined JSON of both "repo" jsons (which gets created right now in backrest - as it is visible) and plan jsons and settings json. A combo or maybe separate jsons in a folder. And I can import it when I want to do that on a new system (personal computer or a server) - maybe I can tweak it if I want. Something like that. I am sharing a small screenshot from [Vorta](https://vorta.borgbase.com/usage/settings/) and maybe it will give a rough idea of what I am talking about. 
<img width="406" alt="Screenshot 2024-09-21 at 3 19 53 PM" src="https://github.com/user-attachments/assets/d7d3ba6e-68b4-4d33-b8a9-00b35512d148">

Basically a "portable backrest setup".

I can open a feature request if you think something like this sits well with your vision for backrest.

> Paths and exclusions

I think I will have to go ahead and add like this and I hope this will be fine 
<img width="722" alt="Screenshot 2024-09-21 at 4 49 40 PM" src="https://github.com/user-attachments/assets/d0213232-c9c7-405f-a44a-4808dd439368">


Thanks for the answers. Really appreciate it.

**Reply by amarendra** [reply]
Yeah this worked.

**Reply by arminus** [reply]
> 1. I think what you're looking at re: using an exclude list as a file is a good approach.

This is also what I'd be looking for.

I have quite some extensive exclude lists for full machine backups (e.g. like this https://github.com/rubo77/rsync-homedir-excludes) for my so far CLI based restic jobs. Adding those patterns one by one through the UI is just not feasible. I can always patch them into the config.json but being able to either reference existing exclude/include files or import them once would really be nice.

**Reply by amarendra** [reply]
Yeah. Include and exclude via new-line separated list files (just like restic supports it) is kinda a must have. I was surprised not to find it here. Or I guess it can be exposed via a flag really.

**Reply by garethgeorge** [reply]
I don’t quite follow — this is absolutely supported and is why backrest allows you to forward flags. 
 
Use the —exclude-file=your exclude list option supported by restic. 

You are not required at all to specify your excludes in the UI’s list.

**Reply by amarendra** [reply]
So you are saying these are supported?

	--files-from $RESTIC_INCLUDE_FILE 
	--exclude-file=$RESTIC_INCLUDE_FILE

as flags in the Plan?

I will set `RESTIC_INCLUDE_FILE` and `RESTIC_INCLUDE_FILE` path value in env vars of the repo. Right? Or just directly pass the file path in flags?

I tried it didn't work. (Maybe I was not supposed to pass the "="

**Reply by amarendra** [reply]
Well, it seems it works @arminus 

> --files-from=direct-file-path
> --exclude-file=direct-file-path

**Reply by arminus** [reply]
Great, I was kind of confused here!
